### PR TITLE
safekeeper: consider partial uploads when pulling timeline

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem:
The control file contains the id of the safekeeper that uploaded it.
Previously, when sending a snapshot of the control file to another sk,
it would eventually be gc-ed by the receiving sk. This is incorrect
because the original sk might still need it later.

## Summary of Changes:
When sending a snapshot and the control file contains an uploaded
segment:
* Create a copy of the segment in s3 with the destination sk in the
  object name
* Tweak the streamed control file to point to the object create in the
  previous step

Note that the snapshot endpoint now has to know the id of the requestor,
so the api has been extended to include the node if of the destination
sk.

Closes https://github.com/neondatabase/neon/issues/8542

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: safekeeper/src/http/routes.rs [hunk 1]
```diff
@@ -565,7 +572,7 @@ pub fn make_router(conf: SafeKeeperConf) -> RouterBuilder<hyper::Body, ApiError>
             request_span(r, tenant_delete_handler)
         })
         .get(
-            "/v1/tenant/:tenant_id/timeline/:timeline_id/snapshot",
+            "/v1/tenant/:tenant_id/timeline/:timeline_id/snapshot/:destination_id",
```
Comment (VladLazar) [inline safekeeper/src/http/routes.rs:575]: @arssher is there any callers of this API apart from the peer SKs?

File: test_runner/regress/test_wal_acceptor.py [hunk 2]
```diff
@@ -2304,3 +2322,131 @@ def test_s3_eviction(
...
+
+    log.info(
+        f"Uploaded segments after post-pull ingest are {src_sk.list_uploaded_segments(tenant_id, timeline_id)}"
+    )
+
+    # Allow for some gc iterations to happen and assert that the original
+    # uploaded partial segment remains in place.
+    time.sleep(5)
+    segs = src_sk.list_uploaded_segments(tenant_id, timeline_id)
+    assert source_partial_segment in segs
```
Comment (VladLazar) [inline test_runner/regress/test_wal_acceptor.py:2433]: Reviewer note: all assertions from this point fail without the fix

File: safekeeper/src/http/routes.rs (hunk 1 above)
Comment (arssher) [reply safekeeper/src/http/routes.rs:575]: no

File: safekeeper/src/pull_timeline.rs [hunk 3]
```diff
@@ -158,13 +166,51 @@ impl WalResidentTimeline {
...
+            .await?;
+
+            scopeguard::defer! {
+                if let Err(err) = control_file::FileStorage::remove_temp(self.get_timeline_dir()) {
+                    tracing::warn!("Failed to remove temp control file: {err}");
+                }
+            }
+
+            let mut control_temp =
+                control_file::FileStorage::write_temp(&control_store, self.get_timeline_dir())
```
Comment (petuhovskiy) [inline safekeeper/src/pull_timeline.rs:204]: What do you think about not using `write_temp` to create a temp file on disk and instead try to write bytes from `write_to_buf` directly to archive?

https://docs.rs/tokio-tar/latest/tokio_tar/struct.Builder.html#method.append_data

I looked briefly at tokio-tar and I see that API for this is not great, but writing a temp file to disk feels too hacky.

Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs:204]: > What do you think about not using write_temp to create a temp file on disk and instead try to write bytes from write_to_buf directly to archive?

Yeah, that's what I tried initially, but couldn't figure out how `Header` should look like. I can look into it again

Comment (petuhovskiy) [reply safekeeper/src/pull_timeline.rs:204]: They have this snippet in the doc:
```
let mut header = Header::new_gnu();
header.set_size(4);
header.set_cksum();

let mut data: &[u8] = &[1, 2, 3, 4];

let mut ar = Builder::new(Vec::new());
ar.append_data(&mut header, "really/long/path/to/foo", data).await?;
let data = ar.into_inner().await?;
```

Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs]: I wasn't sure about file perms, but I think it's fine. Done in https://github.com/neondatabase/neon/pull/8628/commits/fb88c56277d47a96ced3a88c4242e1aa3507765e

File: safekeeper/src/wal_backup_partial.rs [hunk 4]
```diff
@@ -90,6 +96,33 @@ impl State {
...
+        &mut self,
+        source: NodeId,
+        destination: NodeId,
+    ) -> Option<ReplaceUploadedSegment> {
+        let current = self
+            .segments
+            .iter_mut()
+            .find(|seg| seg.status == UploadStatus::Uploaded)?;
+
+        let previous = current.clone();
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply safekeeper/src/wal_backup_partial.rs:132]: Very reasonable request: https://github.com/neondatabase/neon/pull/8628/commits/8ab663f9b9e142a9bd647d9fb69529b85ef03338

File: safekeeper/src/pull_timeline.rs [hunk 5]
```diff
@@ -158,13 +166,47 @@ impl WalResidentTimeline {
...
+                .with_context(|| "failed to serialize control store")?;
+            let mut header = Header::new_gnu();
+            header.set_size(buf.len().try_into().expect("never breaches u64"));
+            ar.append_data(&mut header, CONTROL_FILE_NAME, buf.as_slice())
+                .await
+                .with_context(|| "failed to append to archive")?;
+        } else {
+            let cf_path = self.get_timeline_dir().join(CONTROL_FILE_NAME);
+            let mut cf = File::open(cf_path).await?;
+            ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs]: I think this makes sense. From what I can tell, the persisted and in-mem control file should always be in sync.
Done in https://github.com/neondatabase/neon/pull/8628/commits/9d4ec8625de1861a001926bd18cba5f5358935d9

File: safekeeper/src/pull_timeline.rs [hunk 6]
```diff
@@ -158,13 +166,47 @@ impl WalResidentTimeline {
...
+        destination: NodeId,
     ) -> Result<SnapshotContext> {
         let mut shared_state = self.write_shared_state().await;
         let wal_seg_size = shared_state.get_wal_seg_size();
 
         let cf_path = self.get_timeline_dir().join(CONTROL_FILE_NAME);
-        let mut cf = File::open(cf_path).await?;
-        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+
+        let mut control_store = control_file::FileStorage::load_control_file(&cf_path)?;
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs]: Done in https://github.com/neondatabase/neon/pull/8628/commits/9d4ec8625de1861a001926bd18cba5f5358935d9

File: safekeeper/src/pull_timeline.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs]: `TimelineState::inmem` is public, but I don't think we mutate through it anywhere.

File: safekeeper/src/pull_timeline.rs [hunk 7]
```diff
@@ -158,13 +166,47 @@ impl WalResidentTimeline {
...
+        if let Some(replace) = replace {
+            // The deserialized control file has an uploaded partial. We upload a copy
+            // of it to object storage for the destination safekeeper and send an updated
+            // control file in the snapshot.
+            tracing::info!(
+                "Replacing uploaded partial segment in in-mem control file: {replace:?}"
+            );
+
+            let remote_timeline_path = wal_backup::remote_timeline_path(&self.tli.ttid)?;
+            wal_backup::copy_partial_segment(
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply safekeeper/src/pull_timeline.rs:185]: Hmm, I missed `self.write_shared_state().await;` in the beginning of the function, probably parallel activity can't interfere here.

File: safekeeper/src/pull_timeline.rs [hunk 8]
```diff
@@ -158,13 +162,41 @@ impl WalResidentTimeline {
...
         let mut shared_state = self.write_shared_state().await;
         let wal_seg_size = shared_state.get_wal_seg_size();
 
-        let cf_path = self.get_timeline_dir().join(CONTROL_FILE_NAME);
-        let mut cf = File::open(cf_path).await?;
-        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+        let mut control_store = TimelinePersistentState::clone(shared_state.sk.state());
+        let replace = control_store
+            .partial_backup
+            .replace_uploaded_segment(source, destination)?;
```
Comment (petuhovskiy) [inline safekeeper/src/pull_timeline.rs:174]: It wasn't obvious for me that this function modifies `.partial_backup` in-place, probably worth a comment.

Another option is to just explicitly reassign the state:
```
control_store.partial_backup = wal_backup_partial::State {
  segments: vec![replace.current],
};
```

And then you don't need a reference.

**Review by petuhovskiy — APPROVED** [review]

File: safekeeper/src/pull_timeline.rs (hunk 8 above)
Comment (VladLazar) [reply safekeeper/src/pull_timeline.rs:176]: Added a comment in https://github.com/neondatabase/neon/pull/8628/commits/f423f9f7d8d89bb4e5753d3f5f661372abb71a94

**Review by arssher — APPROVED** [review]

**Comment by arssher** [comment]
Thanks! LGTM.

File: test_runner/regress/test_wal_acceptor.py [hunk 9]
```diff
@@ -2304,3 +2322,142 @@ def test_s3_eviction(
...
+        first_segment_name = "000000010000000000000001"
+        segs = src_sk.list_uploaded_segments(tenant_id, timeline_id)
+
+        candidate_seg = None
+        for seg in segs:
+            if "partial" in seg and "sk1" in seg and not seg.startswith(first_segment_name):
+                candidate_seg = seg
+
+        if candidate_seg is not None:
+            # The term might change, causing the segment to be gc-ed shortly after,
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply test_runner/regress/test_wal_acceptor.py:2379]: Yeah, I didn't get it either, but the test did fail in CI due to a term change, hence the extra check.

File: test_runner/regress/test_wal_acceptor.py [hunk 10]
```diff
@@ -2304,3 +2322,142 @@ def test_s3_eviction(
...
+
+    src_flush_lsn = src_sk.get_flush_lsn(tenant_id, timeline_id)
+    log.info(f"flush_lsn on src before pull_timeline: {src_flush_lsn}")
+
+    pageserver_conn_options = {"password": env.auth_keys.generate_tenant_token(tenant_id)}
+    wait_lsn_force_checkpoint_at(
+        src_flush_lsn, tenant_id, timeline_id, env.pageserver, pageserver_conn_options
+    )
+
+    pt_handle = PropagatingThread(
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply test_runner/regress/test_wal_acceptor.py]: It is indeed: https://github.com/neondatabase/neon/pull/8628/commits/8f88330abf2ee7af7551d571b655644645816860

**Pull request merged**